docs/source/basics/custom_components.rst

```diff
@@ -0,0 +1,171 @@
+        # Config will be parsed from CLI, don't need to pass in here
+        main()
+
+Launching with custom components
+--------------------------------
+torchtune supports full experimentation with custom models, datasets, optimizers, or any fine-tuning component. You can define
+these locally in your repo and use them in your recipes and configs that you can launch with ``tune run``.
+
+We recommend following the "builder" pattern when making your components. This means creating "builder" functions that set up
+the classes you need with a few high level parameters that can be modified easily from the config. For example, we can define custom
+model and dataset builders in our project directory:
+
+.. code-block:: python
+
+    #
+    # In models/custom_decoder.py
+    #
+    class CustomTransformerDecoder(nn.Module):
+        # A custom architecture not present in torchtune
+
+    # Builder function for the custom model
+    def custom_model(num_layers: int, classification_head: bool = False):
+        # Any setup for defining the class
+        ...
+        # Return the module you want to train
+        return CustomTransformerDecoder(...)
+
+.. code-block:: python
+
+    #
+    # In datasets/custom_dataset.py
+    #
+    from torchtune.datasets import SFTDataset, PackedDataset
+    from torchtune.data import InputOutputToMessages
+    from torchtune.modules.tokenizers import ModelTokenizer
+
+    # Example builder function for a custom code instruct dataset not in torchtune, but using
+    # different dataset building blocks from torchtune
+    def tiny_codes(tokenizer: ModelTokenizer, packed: bool = True):
+        """
+        Python subset of nampdn-ai/tiny-codes. Instruct and code response pairs.
+        """
+        ds = SFTDataset(
+            model_transform=tokenizer,
+            source="nampdn-ai/tiny-codes",
+            message_transform=InputOutputToMessages(
+                column_map={"input": "prompt", "output": "response"},
+            ),
+            filter_fn=lambda x: x["language"] == "python",
+            split="train",
+        )
+        if packed:
+            return PackedDataset(ds, max_seq_len=tokenizer.max_seq_len, split_across_pack=False)
+        else:
+            return ds
+
+.. note::
+
+    If you are using a default torchtune recipe with a custom dataset, you must define the first
+    positional argument to be the tokenizer or model transform. These are automatically passed into
+    dataset instantiation and are defined separately in the config, not under the dataset field.
+
+You can define the custom model and custom dataset in the config using the relative import path from where
+you are launching with ``tune run``. It is best to define the path relative to your project root directory
+and launch from there.
+
+.. code-block:: yaml
+
+    # In YAML file config/custom_finetune.yaml
+    model:
+      _component_: models.custom_decoder.custom_model
+      num_layers: 32
+
```

Comment:
```suggestion
      num_layers: 32
      classification_head: False

```

Reply:
just to be super obvious what we're doing here

Reply:
wanted to subtly show that this is an optional param so didn't need to specify in config. but I'm ok either way

Reply:
```
# this is an optional param, so you can also omit this from the config
```
mayhaps